samples/infix/writers.cc

```diff
@@ -7,20 +13,45 @@ namespace
 
   // clang-format off
   inline const auto wf_pass_maths = 
-    infix::wf 
-    | (Assign <<= Ident * Literal) 
-    | (Output <<= String * Literal) 
-    | (Literal <<= wf_literal)
+    infix::wf
+    // The Op >>= prefix ensures the internal structure of the WF is set up properly.
+    // The disjunction cannot otherwise be nested inside *
+    // Technically it names the disjunction "Op", but we do not use that name.
+    | (Assign <<= Ident * (Op >>= Literal | Expression)) 
```

Comment:
Error logic was moved to a separate pass because it was triggering on partial evaluations, whenever rule evaluation went outside `once`.

Your suggestion is interesting; it does make the WF definition smaller. It also means I have to go edit all of the passes in `README.md` which list rules in the previous format; I would probably have to rewrite parts of the prose around them too. Not doing that was originally to keep the PR diff a bit smaller.

Well, maybe that's what has to happen.

It is unsatisfying in any case that I can't just replace a token type in a WF (while preserving structure like children etc). That would negate all of this, and let us write simpler rules.

Reply:
I tried it, complete with a relatively large .md diff in both README and beyond-infix.

Reply:
Probably fine now. Feel free to "re-open" conversation if the committed result seems weird.